File: src/main/php/g0601_0700/s0647_palindromic_substrings/Solution.php
<?php

namespace leetcode\g0601_0700\s0647_palindromic_substrings;

// #Medium #Top_100_Liked_Questions #String #Dynamic_Programming #Big_O_Time_O(n^2)_Space_O(n)
// #2023_12_24_Time_25_ms_(75.00%)_Space_19.6_MB_(50.00%)

class Solution {
    private function expand($a, $l, $r, &$res) {
        while ($l >= 0 && $r < count($a)) {
            if ($a[$l] != $a[$r]) {
                return;
            } else {
                $res[0]++;
                $l--;
                $r++;
            }
        }
    }

    /**
     * @param String $s
     * @return Integer
     */
    public function countSubstrings($s) {
        $a = str_split($s);
        $res = array(0);
        for ($i = 0; $i < count($a); $i++) {
            $this->expand($a, $i, $i, $res);
            $this->expand($a, $i, $i + 1, $res);
        }
        return $res[0];
    }
}

File: src/main/go/g0601_0700/s0647_palindromic_substrings/solution.go
package s0647_palindromic_substrings

// #Medium #String #Dynamic_Programming #Big_O_Time_O(n^2)_Space_O(n)
// #2025_05_11_Time_0_ms_(100.00%)_Space_4.02_MB_(35.44%)

func expand(a []rune, l int, r int, res *int) {
	for l >= 0 && r < len(a) {
		if a[l] != a[r] {
			return
		} else {
			*res++
			l--
			r++
		}
	}
}

func countSubstrings(s string) int {
	a := []rune(s)
	res := 0
	for i := 0; i < len(a); i++ {
		expand(a, i, i, &res)
		expand(a, i, i+1, &res)
	}
	return res
}
